File: cole/day14_balanced-binary-tree/balanced-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

trait BalancedTree {
    fn walk(&mut self) -> i32;
    fn balanced(&mut self) -> bool;
}

impl BalancedTree for TreeNode {

    fn walk(&mut self) -> i32 {
        let l = self.left.as_mut().map(|node| node.borrow_mut().walk()).unwrap_or(0);
        let r = self.right.as_mut().map(|node| node.borrow_mut().walk()).unwrap_or(0);
        match (l - r).abs() > 1  || l < 0 || r < 0 {
            true => -1,
            false => 1 + i32::max(l, r)
        }
    }
    fn balanced(&mut self) -> bool {
        self.walk() > 0
    }
}

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        root.map(|node| node.borrow_mut().balanced()).unwrap_or(true)
    }
}

File: jagraj/14/balancedBinaryTree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def height(self, root):
        return 0 if root is None else 1 + max(self.height(root.left), self.height(root.right))

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return True if root is None else abs(self.height(root.left) - self.height(root.right)) <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right)
